File: oim-jruby/update_users_mt.rb
require 'java'
require 'xlclient'

include_class('java.lang.Exception') {|package,name| "J#{name}" }
include_class('java.lang.Thread') {|package,name| "J#{name}" }
include_class 'java.lang.System' 
include_class 'java.util.HashMap'
include_class('Thor.API.tcUtilityFactory') {|package,name| "OIM#{name}"}


class Worker < JThread
    attr_accessor :xlclient, :tid, :count

    def initialize
    end

    def setVars(xlclient, tid, begusr, endusr)
        @xlclient = xlclient
	@tid = tid
	@begusr = begusr
	@endusr = endusr
    end

    def run
        puts "Running #{@tid}"
        usrIntf = @xlclient.getUtility('usr')
	    filter = ['Users.Key', 'Users.User ID', 
	    'Users.Row Version'].to_java(:String)

	    for usr_key in (@begusr..@endusr)
	        usrMap = HashMap.new({
		    'Users.Key' => usr_key.to_s
	        })

	        rs = usrIntf.findUsersFiltered(usrMap, filter)
	        userID = rs.getStringValue('Users.User ID')

	        updateMap = HashMap.new({
		        'Users.First Name' => userID + '555',
		        'Users.Last Name' => userID + '666'
	        })

	        t1 = System.currentTimeMillis
	        usrIntf.updateUser(rs, updateMap)
	        t2 = System.currentTimeMillis
	        delta = t2-t1
	        puts "Thread #{@tid} updated user with key = #{usr_key} time = #{delta}"
	    end
    end
end

# actual script
begusr = 11
step = 500
numThreads = 20

xlclient = XLAPIClient.new
xlclient.defaultLogin

threads = []

numThreads.times do |i|
    w = Worker.new
    x = i * step + begusr
    y = x + step
    w.setVars(xlclient, 'Thread' + i.to_s, x, y)
    threads.push(w)
end

threads.each do |t| 
    puts "Starting #{t.tid}"
    t.start
end

threads.each do |t| 
    puts "Joining #{t.tid}"
    t.join
end

xlclient.close
System.exit 0

File: oim-jython/updateUsersMT.py
from xlclient import *
from java.util import *
from java.lang import *

import sys

class T(Thread):
    def __init__(self, xlclient, tid, x, y):
	self.xlclient = xlclient
	self.tid = tid
	self.x = x
	self.y = y

    def __str__(self):
	return self.tid
    
    def run(self):
	usrIntf = self.xlclient.usrIntf
	usrCols = ['Users.Key', 'Users.Row Version', 'Users.User ID']

	usrMap = HashMap()

	for i in range(self.x, self.y):
	    usrMap.put('Users.Key', str(i))

	    rs = usrIntf.findUsersFiltered(usrMap, usrCols)

	    updateMap = HashMap()
	    updateMap.put('Users.First Name', rs.getStringValue('Users.User ID') + 'Fst')
	    updateMap.put('Users.Last Name', rs.getStringValue('Users.User ID') + 'Lst')
	    updateMap.put('USR_UDF_ENCUDF1', 'tt')
	    updateMap.put('USR_UDF_ENCUDF2', 'tt')
	    updateMap.put('USR_UDF_PASSWDUDF', 'tt')

	    t1 = System.currentTimeMillis()
	    usrIntf.updateUser(rs, updateMap)
	    t2 = System.currentTimeMillis()
	    delta = t2-t1
	
	    print 'Updating user ', rs.getStringValue('Users.User ID'), ' time = ', str(delta)
	    
	
if __name__ == '__main__':
    xlclient = XLAPIClient('xelsysadm', 'xelsysadm')
    numThreads = 21
    step = 5000
    threads = []

    for i in range(numThreads):
	x = i * step + 1
	y = x + step
	t = T(xlclient, 'Thread' + str(i), x, y)
	threads.append(t)

    for t in threads:
	print 'Running ', t.tid
	t.start()

    for t in threads:
	print 'Joining ', t.tid
	t.join()

    xlclient.close()
    sys.exit(0)
